File: TTK4145 - Sanntidsprogrammering/Exercises/Ex01-Intro/main.go
package main



import (
    . "fmt"
    "runtime"
)

var i int = 0


func threadFunction_1(message chan int, done_message_1 chan bool){
	for k := 0; k < 1000002; k++{
		i := <- message
		i ++
		message <- i
	}
	done_message_1 <- true
}


func threadFunction_2(message chan int, done_message_2 chan bool){
	for k := 0; k < 1000000; k++{
		i := <- message
		i --
		message <- i
	}
	done_message_2 <- true
}


func main(){

	message := make(chan int, 1)

	done_message := make(chan bool)

	runtime.GOMAXPROCS(runtime.NumCPU())

	message <- 0
	go threadFunction_1(message, done_message)
	go threadFunction_2(message, done_message)

	<-done_message
	<-done_message

	i := <- message
	Println("Value after both threads running, i = ", i)
}

File: TTK4145 - Sanntidsprogrammering/Exercises/Ex01-Intro/main.c
#include <pthread.h>
#include <stdio.h>

int i = 0;

pthread_mutex_t lock;

void* threadFunction_1(){
	for(int k = 0; k < 1000003; k++){
		pthread_mutex_lock(&lock);
		i++;
		pthread_mutex_unlock(&lock);
	}

}

void* threadFunction_2(){
	for(int k = 0; k < 1000000; k++){
		pthread_mutex_lock(&lock);
		i--;
		pthread_mutex_unlock(&lock);
	}
}

int main(){
	pthread_t thread_1;
	pthread_t thread_2;

	pthread_create(&thread_1, NULL, threadFunction_1, NULL);
	pthread_create(&thread_2, NULL, threadFunction_2, NULL);

	pthread_join(thread_1, NULL);
	pthread_join(thread_2, NULL);
	printf("EX 2 After running both threads, i = %d\n", i);
	
	return 0;
}